edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/upstream/BrokeredCloudProxyDispatcher.cs

```diff
@@ -0,0 +1,509 @@
+                        else
+                        {
+                            Events.CannotFindGuid(ackedGuid);
+                        }
+                    }
+
+                    break;
+
+                default:
+                    Events.UnknownAckType(cmd, ackedGuid);
+                    break;
+            }
+        }
+
+        void HandleTwinResponse(string id, string res, string rid, byte[] payload)
+        {
+            var messageBuilder = new EdgeMessage.Builder(payload);
+            messageBuilder.SetSystemProperties(
+                new Dictionary<string, string>()
+                {
+                    [SystemProperties.StatusCode] = res,
+                    [SystemProperties.CorrelationId] = rid,
+                });
+
+            var message = messageBuilder.Build();
+
+            long ridAsLong;
+            if (!long.TryParse(rid, out ridAsLong)) // TODO should use as string itself to make less sensitive to the actual format?
+            {
+                Events.CannotParseRid(rid);
+                return;
+            }
+
+            if (!this.pendingTwinRequests.TryRemove(ridAsLong, out var tsc))
+            {
+                Events.CannotFindRid(rid);
+                return;
+            }
+
+            tsc.SetResult(message);
+
+            Events.TwinResultReceived(id, ridAsLong);
+        }
+
+        void HandleDesiredProperyUpdate(string id, string version, byte[] payload)
+        {
+            var messageBuilder = new EdgeMessage.Builder(payload);
+            messageBuilder.SetSystemProperties(
+                new Dictionary<string, string>()
+                {
+                    [SystemProperties.Version] = version
+                });
+
+            _ = this.edgeHub.UpdateDesiredPropertiesAsync(id, messageBuilder.Build());
+        }
+
+        void HandleDirectMethodCall(string id, string method, string rid, byte[] payload)
+        {
+            // TODO acquire the response timeout from the message
+            var callingTask = this.edgeHub.InvokeMethodAsync(rid, new DirectMethodRequest(id, method, payload, TimeSpan.FromMinutes(1)));
+            callingTask.ContinueWith(
+                    async response =>
+                    {
+                        var status = response.IsCompletedSuccessfully ? response.Result.Status : 500; // TODO check the correct status code to return
+                        var topic = string.Format(DirectMethodResponseTemplate, id, response.Result.Status, rid);
+
+                        await this.SendUpstreamMessageAsync(RpcCmdPub, topic, response.Result.Data ?? this.emptyArray);
+                    });
+        }
+
+        string GetIdFromMatch(Match match)
+        {
+            return match.Groups["id2"].Success
+                            ? $"{match.Groups["id1"].Value}/{match.Groups["id2"].Value}"
+                            : match.Groups["id1"].Value;
+        }
+
+        async Task WaitCompleted(Task task, string id, long rid)
```

Comment:
TaskEx in Microsoft.Azure.Devices.Edge.Util has a TimeoutAfter method that can be used here (about the same with a Delay)